aws_signing_helper/pkcs11_signer.go

```diff
@@ -334,49 +649,168 @@ func checkPrivateKeyMatchesCert(module *pkcs11.Ctx, session pkcs11.SessionHandle
+		pinPkcs11, _ := cert_uri.GetQueryAttribute("pin-value", false)
+		session, slot_nr, logged_in, cert_obj, err = getMatchingCerts(module, slots, certIdentifier, cert_uri, true, &pinPkcs11)
 		if err != nil {
 			goto fail
 		}
+		certificate = cert_obj[0].x509
 	}
 
-	if pinPkcs11 == "-" {
-		fmt.Fprintln(os.Stderr, "Please enter your user pin:")
-		pinPkcs11Bytes, err = term.ReadPassword(int(syscall.Stdin))
-		if err != nil {
-			err = errors.New("unable to read PKCS#11 user pin")
+	// If no explicit private-key option was given, use it. Otherwise
+	// we look in the same place as the certificate URI as directed by
+	// http://david.woodhou.se/draft-woodhouse-cert-best-practice.html#rfc.section.8.2
+	if (privateKeyId != "") {
+		key_uri = pkcs11uri.New()
+		err = key_uri.Parse(privateKeyId)
+		if (err != nil) {
 			goto fail
 		}
-
-		pinPkcs11 = string(pinPkcs11Bytes[:])
-		strings.Replace(pinPkcs11, "\r", "", -1) // Remove CR
+	} else {
+		key_uri = cert_uri;
 	}
+	pinPkcs11, _ = key_uri.GetQueryAttribute("pin-value", false)
 
-	err = module.Login(session, pkcs11.CKU_USER, pinPkcs11)
-	if err != nil {
+	// This time we're looking for a *single* slot, as we (presumably)
+	// will have to log in to access the key.
+	slots = matchSlots(slots, key_uri)
+	if len(slots) == 1 {
+		if slot_nr != slots[0].id {
+			if session != 0 {
+				module.CloseSession(session)
+			}
+			session = 0
+			logged_in = false
+			slot_nr = slots[0].id
+		}
+	} else {
+		// If the URI matched multiple slots *but* one of them is the
+		// one (slot_nr) that the certificate was found in, then use
+		// that.
+		for _, slot := range slots {
+			if slot.id == slot_nr {
+				goto got_slot
+			}
+		}
+		err = errors.New("Could not identify unique slot for PKCS#11 key")
 		goto fail
 	}
 
-	templatePrivateKey = []*pkcs11.Attribute{
-		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PRIVATE_KEY),
+got_slot:
+	if session == 0 {
+		session, err = module.OpenSession(slot_nr, pkcs11.CKF_SERIAL_SESSION|pkcs11.CKS_RO_PUBLIC_SESSION)
+		if err != nil {
+			goto fail
+		}
 	}
+
+	if !logged_in {
+		// And *now* we fall back to prompting the user for a PIN.
+		// Perhaps we should do this via a callback function, and
+		// do it whenever we attempt logins? And loop on failure
+		// in case the user mistyped it?
+		if pinPkcs11 == "" {
+			fmt.Fprintln(os.Stderr, "Please enter your user pin:")
+			pinPkcs11Bytes, err = term.ReadPassword(int(syscall.Stdin))
+			if err != nil {
+				err = errors.New("unable to read PKCS#11 user pin")
+				goto fail
+			}
+
+			pinPkcs11 = string(pinPkcs11Bytes[:])
+			strings.Replace(pinPkcs11, "\r", "", -1) // Remove CR
+		}
+
+		err = module.Login(session, pkcs11.CKU_USER, pinPkcs11)
+		if err != nil {
+			goto fail
+		}
+	}
+retry_search:
+	templatePrivateKey = getFindTemplate(key_uri, pkcs11.CKO_PRIVATE_KEY)
+
 	if err = module.FindObjectsInit(session, templatePrivateKey); err != nil {
 		goto fail
 	}
```

Comment:
This looks like code that I wrote, but maybe line 817 has to be done in a loop to enumerate all objects within the session? I think that's what I did for the certificates.